tiledb/sm/filter/filter.cc

```diff
@@ -106,28 +117,38 @@ Status Filter::set_option(FilterOption option, const void* value) {
   return set_option_impl(option, value);
 }
 
-Status Filter::deserialize(ConstBuffer* buff, Filter** filter) {
+std::tuple<Status, std::optional<std::shared_ptr<Filter>>> Filter::deserialize(
+    ConstBuffer* buff) {
+  Status st;
   uint8_t type;
-  RETURN_NOT_OK(buff->read(&type, sizeof(uint8_t)));
+  st = buff->read(&type, sizeof(uint8_t));
+  if (!st.ok()) {
+    return {st, std::nullopt};
+  }
   uint32_t filter_metadata_len;
-  RETURN_NOT_OK(buff->read(&filter_metadata_len, sizeof(uint32_t)));
+  st = buff->read(&filter_metadata_len, sizeof(uint32_t));
+  if (!st.ok()) {
+    return {st, std::nullopt};
+  }
 
-  auto* f = create(static_cast<FilterType>(type));
-  if (f == nullptr)
-    return LOG_STATUS(Status::FilterError("Deserialization error."));
+  auto&& [st_filter, filter]{create(static_cast<FilterType>(type))};
+  if (!st_filter.ok()) {
+    return {st_filter, std::nullopt};
+  }
 
   auto offset = buff->offset();
-  RETURN_NOT_OK_ELSE(f->deserialize_impl(buff), tdb_delete(f));
+  st = filter.value()->deserialize_impl(buff);
```

Comment:
`deserialize_impl` can't be made C.41 compliant. We can't use it to accomplish the goals of this work.

We can use non-member functions such as `deserialize_bit_width_reduction_filter(...)`, for example, with another `switch` on filter type.